src/java/org/apache/cassandra/locator/ReplicaPlan.java

```diff
@@ -237,4 +237,17 @@ public int requiredParticipants()
         }
     }
 
+    /**
+     * Used by AbstractReadExecutor, {Data,Digest}Resolver and ReadRepair to share a ReplicaPlan whose 'contact' replicas
+     * we progressively modify via various forms of speculation (initial speculation, rr-read and rr-write)
+     * @param <P>
+     */
+    public static class Shared<P extends ReplicaPlan<?, ?, ?>>
```

Comment:
This is kind of a weird place it end up in terms of mutability and immutability. It's not thread safe and if it's supposed to be always referencing the latest instance why immutable?

Reply:
I'm not sure I really understand the question or criticism.  Could it be mutable?  For starters, it would require a lot of other changes to ReplicaPlan and ReplicaCollection, that don't seem warranted for this single use case.  Besides that, yes, but this would be less threadsafe (this version is actually perfectly safe to have multiple readers, as we do, they just may not see the latest 'contacts' value promptly).  I'd be happy to upgrade it to fully threadsafe if it would make us happier, but this was changed to non-threadsafe specifically in response to prior feedback about the old code that it unnecessarily used volatile.  It would be trivial to make it fully threadsafe, and the cost would actually be fairly low, and while I'm fairly sure it isn't needed, it would be easier to reason about.